crates/hotshot/src/types/handle.rs

```diff
@@ -22,6 +29,65 @@ use tracing::instrument;
 
 use crate::{traits::NodeImplementation, types::Event, Memberships, SystemContext};
 
+/// A signal for coordinating shutdown across multiple tasks.
+///
+/// This struct provides methods to initiate a shutdown, check the shutdown status,
+/// and wait for a shutdown signal.
+pub struct ShutdownSignal {
```

Comment:
I am hesitant about the introduction of a bespoke shutdown signal since we have something like this already in the codebase. Perhaps @ss-es can give context on the byzantine tests specifically, but it seems like we might be able to somehow re-use that behavior? Especially considering that this type is bolted onto the SYstemContextHandle anyway.

Reply:
@shamb0 I think this looks good overall, though I agree with Jarred about the extra shutdown signal.

I think the preferred way to do this would be to listen on the `internal_event_stream` for a `HotShotEvent::Shutdown` (which is the shutdown signal for all tasks). I don't think the network message task has an internal event stream receiver, but you can pass it to the task state on creation in `tasks/mod.rs`.

Reply:
Oh, nevermind -- I see you already made this change! I'll look through the code and add some comments